src/Stackage/Package/Git/Types.hs

```diff
@@ -0,0 +1,211 @@
+import qualified Data.Map as Map
+import Foreign.Storable (peekByteOff)
+import Time.System (timeCurrent)
+
+import qualified Data.Git as G
+
+-- | User that will be used for creating commits (both author and commiter) and
+-- tags.
+data GitUser = GitUser
+  { userName :: ByteString
+    -- ^ Full name
+  , userEmail :: ByteString
+    -- ^ Email
+  , userGpgKey :: Maybe String
+    -- ^ GPG public key.
+  } deriving (Show)
+
+
+-- | Information about the git repository.
+data GitInfo = GitInfo
+  { gitAddress :: String
+    -- ^ Git address of the repository were it can be cloned from using SSH key.
+  , gitBranchName :: String
+    -- ^ Name of the branch that updates should be committed to. (Remote branch
+    -- with the same name is expected to exist).
+  , gitUser :: GitUser
+    -- ^ User information to be used for the commits.
+  , gitTagName :: Maybe ByteString
+    -- ^ Create a tag of a commit.
+  , gitLocalPath :: FilePath
+    -- ^ Where in the file system this repository is located.
+  , gitRootTree :: Maybe (WorkTree ShortRef ShortRef)
+  }
+
+
+data GitInstance = GitInstance
+  { gitRepo :: G.Git
+    -- ^ Git repository instance.
+  , gitBranchRef :: MVar G.Ref
+    -- ^ Reference of the commit, that current branch is pointing to.
+  , gitWorkTree :: MVar (WorkTree ShortRef ShortRef, WorkTree () GitFile)
+    -- ^ Current work tree written to the git store.
+  }
+
+
+-- | Everything that is necessary to operate on a repository.
+data GitRepository = GitRepository
+  { repoInstance :: GitInstance
+  , repoInfo :: GitInfo
+  }
+
+
+
+-- | A file that is prepared for being written into git store.
+data GitFile = GitFile
+  { gitFileRef :: !G.Ref
+    -- ^ SHA1 of the blob.
+  , gitFileType :: !FileType
+    -- ^ One of the few possible file types.
+  , gitFileZipped :: !ByteString
+    -- ^ A blob object: marshalled and compressed file.
+  } deriving (Show)
+
+type GitTree = G.Tree
+
+type GitCommit = G.Commit
+
+type GitTag = G.Tag
+
+-- | A file name as it is expected to look like by git. This data type is
+-- crutial for keeping files properly sorted in the work tree, as it is expected
+-- by git.
+data FileName
+  = FileName !BS.ShortByteString
+    -- ^ Just a file name without a path to it.
+  | DirectoryName !BS.ShortByteString
+    -- ^ Directory name with a trailing slash, without it trees wouldn't get
+    -- sorted properly: @("0.1" > "0.1.1") /= ("0.1/" > "0.1.1/")@
+  deriving (Show)
+
+
+type TreePath = [FileName]
+
+
+-- | Convert file path to a tree path.
+toTreePath :: FilePath -> TreePath
+toTreePath path = foldr toFileName [] splitPath'
+  where
+    splitPath' = S8.split '/' $ S8.pack path
+    toFileName fName []
+      | null fName =
+        error $ "Tree path should end with a file name, not a directory: " ++ path
+      | otherwise = [FileName $ BS.toShort fName]
+    toFileName fDir tp = DirectoryName (BS.toShort $ S8.snoc fDir '/') : tp
+
+
+instance Eq FileName where
```

Comment:
I'm about 99% certain I understood that comment above, and that my suggested `normalize` function _would_ work. If not, then this comment needs a lot more explanation.

```haskell
normalize :: FileName -> ByteString
normalize (FileName x) = BS.fromShort x
normalize (DirectoryName x) = S8.init $ BS.fromShort x
```

There's a separate but related question about whether these kinds of instances are really appropriate at all, I'm not convinced that that's the case here, but I'm willing to leave it be. But it definitely needs to be commented: having an `Eq` instance where `x == y` but `x` and `y` are observably different is very confusing.